backend/src/lib.rs

```diff
@@ -184,91 +75,40 @@ pub type Proof = Vec<u8>;
     module, wrapping the traits implemented by each backend.
 */
 
-/// Dynamic interface for a backend.
-pub trait Backend<F: FieldElement> {
-    fn add_verification_key(
-        &mut self,
-        pil: &Analyzed<F>,
-        fixed: &[(String, Vec<F>)],
-        vkey: Vec<u8>,
-    );
-
-    fn verification_key(
+/// Dynamic interface for a backend factory.
+pub trait BackendFactory<F: FieldElement> {
+    /// Create a new backend object.
+    fn create<'a>(
         &self,
-        pil: &Analyzed<F>,
-        fixed: &[(String, Vec<F>)],
-    ) -> Result<Vec<u8>, Error>;
+        pil: &'a Analyzed<F>,
+        fixed: &'a [(String, Vec<F>)],
+        setup: Option<&mut dyn io::Read>,
+        verification_key: Option<&mut dyn io::Read>,
+    ) -> Result<Box<dyn Backend<'a, F> + 'a>, Error>;
+
+    /// Generate a new setup.
+    fn generate_setup(&self, size: DegreeType, output: &mut dyn io::Write) -> Result<(), Error>;
```

Comment:
Why is setup the only function that goes into this trait?

Reply:
shouldn't it also go into the Backend trait like the other functionalities?

Reply:
This is because how we use it in `powdr setup` (I think), where we can call it without needing the analyzed pil or the fixed columns. Verifying and proving, on the other hand, do require the full object with the common stuff.

Reply:
I see. Ok... I still find it ugly that a backend that has nothing to do with setup needs to actively implement this function and refuse it, but well.. that's the trade off of having the dynamic interfaces

Reply:
Oh, yeah. I forgot about it, but I was planning to provide default implementation to everything that is optional for a backend to have.

Reply:
Much better now.